### DISCUSSION TITLE
How to combine tabs

### DISCUSSION THREAD

**Discussion by Reiji777** [CoderLine/alphaTab#654] [Q&A] [answered]
UPDATE: After learning the basics of Midi, I somehow managed to make it work. My initial approach causes the two tracks to be in the same midi channel, and that's why the instruments sounded wrong. Fixing the channels solved my issue.

My approach is kinda hacky so I'll try diving into the code for a more proper way.

-------------------------
ORIGINAL:
I have a tab file containing a guitar track, and another tab file containing a bass track. How could I properly combine the two tab files (tracks) so that they can be played together?

So far I tried to load the two tabs individually, and then push the second tab's tracks to the first tab (see pseudo code below for example).
By doing so, the notes of all the tracks can be played correctly, but the first tab's track starts sounding like the instrument of the second tab.

```
const firstScore = api.score; 
//Load the second tab (in AlphaTex)
api.tex(secondScoreTex);
const newScore = api.score;
firstScore.tracks.push(...newScore.tracks)
api.load(firstScore);
```

**Comment by Danielku15** [comment, marked as answer]
Sorry for the late answer, it seems I somehow missed this question completely. Combining the tracks of different files might not be so easy. There are quite some things which are expected to be consistent across the tracks. One main rule is: the number of bars within the staves of a track must be the same as the number of masterbars: https://alphatab.net/docs/reference/score

Here an untested example of how you could organize your code: 

```js
function createEmptyBar(staff) {
    const bar = new Bar();                
    staff.addBar(bar);
    let voiceCount = 1;
    if(bar.index > 0) {
        bar.clef = bar.previousBar.clef;
        voiceCount = bar.previousBar.voices.length;
    }
    
    for(let i = 0; i < voiceCount) {    
        const voice = new Voice();
        bar.addVoice(voice);
        
        const beat = new Beat();
        beat.isEmpty = true;
        voice.addBeat(beat);
    }
    
    return bar;
}

function mergeTrack(target, sourceTrack) {
    // consolidate track
    for(const staff in sourceTrack.staves) {
        if(staff.bars.length > target.masterBars.length) {
            // too many bars, need to truncate
            staff.bars.length = target.masterBars.length;
        } else {
            // too less bars, need to create some empty ones
            while(staff.bars.length < target.masterBars.length) {
                addEmptyBar(staff);
            }
        }
        // fix other things like midi channel
    }
    
    target.addTrack(sourceTrack);
}

function mergeScores(settings, target, source) {
    for(const sourceTrack of source.tracks) {
        mergeTrack(target, sourceTrack); // merge track by track
    }
    target.finish(settings); // important to call to reconsolidate the data model. 
}
```

Or an easier alternative could be to send the scores through the JSON serializer. There you could work more on a simplified JSON base and the serializer takes already care of some consolidation steps. Might cost more CPU though. 

https://alphatab.net/docs/guides/lowlevel-apis#serialize-data-model-fromto-json

**Reply by serenity81** [reply]
Are target and source in mergeScores(settings, target, source) urls to musicxml files?
P.S I am trying to merge tracks from two scores into one score

**Reply by Danielku15** [reply]
No, this logic here is specific to the alphaTab data model.

**Reply by serenity81** [reply]
> No, this logic here is specific to the alphaTab data model.

Its ok I will use alphaTab data model it will work for me this way as well. but how would accomplish the same using the JSON serializer?

**Reply by serenity81** [reply]
> Sorry for the late answer, it seems I somehow missed this question completely. Combining the tracks of different files might not be so easy. There are quite some things which are expected to be consistent across the tracks. One main rule is: the number of bars within the staves of a track must be the same as the number of masterbars: https://alphatab.net/docs/reference/score
> 
> Here an untested example of how you could organize your code:
> 
> ```js
> function createEmptyBar(staff) {
>     const bar = new Bar();                
>     staff.addBar(bar);
>     let voiceCount = 1;
>     if(bar.index > 0) {
>         bar.clef = bar.previousBar.clef;
>         voiceCount = bar.previousBar.voices.length;
>     }
>     
>     for(let i = 0; i < voiceCount) {    
>         const voice = new Voice();
>         bar.addVoice(voice);
>         
>         const beat = new Beat();
>         beat.isEmpty = true;
>         voice.addBeat(beat);
>     }
>     
>     return bar;
> }
> 
> function mergeTrack(target, sourceTrack) {
>     // consolidate track
>     for(const staff in sourceTrack.staves) {
>         if(staff.bars.length > target.masterBars.length) {
>             // too many bars, need to truncate
>             staff.bars.length = target.masterBars.length;
>         } else {
>             // too less bars, need to create some empty ones
>             while(staff.bars.length < target.masterBars.length) {
>                 addEmptyBar(staff);
>             }
>         }
>         // fix other things like midi channel
>     }
>     
>     target.addTrack(sourceTrack);
> }
> 
> function mergeScores(settings, target, source) {
>     for(const sourceTrack of source.tracks) {
>         mergeTrack(target, sourceTrack); // merge track by track
>     }
>     target.finish(settings); // important to call to reconsolidate the data model. 
> }
> ```
> 
> Or an easier alternative could be to send the scores through the JSON serializer. There you could work more on a simplified JSON base and the serializer takes already care of some consolidation steps. Might cost more CPU though.
> 
> https://alphatab.net/docs/guides/lowlevel-apis#serialize-data-model-fromto-json

I think the JS code has issues like looping and getting variable length and such (i know author said its not tested) but i couldn't easily manage to adapt it to my needs. I think it was posted to show top level of how it can be done.

What I managed to do is use the Json serializer to save the two Scores (after loading them in AlphaTab) into a text file then use PHP to handle merging based on the concept of adding empty bars and measuring length suggested in JS code. Yt was easier for me to do it using PHP as I could visually inspect JSON structure changes happening on the spot rather than dealing with alphaTab score model directly. For now I am able to merge tracks from two Scores but not entire sure if timing is 100% only time will tell.

**Reply by Danielku15** [reply]
@serenity81 
My recommendation nowadays is to rather stick with the model based merging most consolidation happens either in the `finish` method or already when adding items (e.g. via `addTrack`)

A very basic working example: 
https://jsfiddle.net/danielku15/bz6ypa4u/

The same on a JSON basis. It still requires the `JsonConverter.jsObjectToScore` to create again a proper consolidated Score (that's why model merging makes more sense, but it works). 
https://jsfiddle.net/danielku15/4fpunxeL/

If you share some more concrete insights on what you plan to do I might have further ideas how you could do things.

**Reply by serenity81** [reply]
Thanks for the reply yes I am using JsonConverter.jsObjectToScore to revert back from JSON to Score Data model. But will try your examples and report back.
